File: src/main/scala/akka/cluster/ddata/replicator/bf/Cell.java
package akka.cluster.ddata.replicator.bf;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

public final class Cell {
    private int count;
    private long idSum;
    private int hashSum;

    public void add(long id, int idHashValue) {
        count++;
        idSum ^= id;
        hashSum ^= idHashValue;
    }

    public Cell() {
    }

    public Cell(int count, long idSum, int hashSum) {
        this.count = count;
        this.idSum = idSum;
        this.hashSum= hashSum;
    }

    public boolean isPure(MessageDigest digestFunction) {
        if ((count == -1 || count == 1)
                && (InvertibleBloomFilter.genIdHash(String.valueOf(idSum).getBytes(StandardCharsets.UTF_8), digestFunction) == hashSum))
            return true;

        return false;
    }

    public void setCount(int count) { this.count = count; }

    public int getCount() {
        return count;
    }

    public void setIdSum(long idSum) { this.idSum = idSum; }

    public long getIdSum() {
        return idSum;
    }

    public void setHashSum(int hashSum) {
        this.hashSum = hashSum;
    }

    public int getHashSum() {
        return hashSum;
    }
}

File: src/main/scala/akka/cluster/ddata/replicator/bfs/Cell.scala
package akka.cluster.ddata.replicator.bfs

import akka.cluster.ddata.replicator.bf.InvertibleBloomFilter

import java.nio.charset.StandardCharsets
import java.security.MessageDigest

final case class Cell(count: Int, idSum: Long, hashSum: Int) { self =>

  def add(id: Int, idHashValue: Int): Cell =
    self.copy(count = count + 1, idSum = idSum ^ id, hashSum = hashSum ^ idHashValue)

  def isPure(digestFunction: MessageDigest): Boolean =
    (count == -1 || count == 1)
      && (InvertibleBloomFilter.genIdHash(
        String.valueOf(idSum).getBytes(StandardCharsets.UTF_8),
        digestFunction
      ) == hashSum)

}
